phonon/update.py

```diff
@@ -234,5 +198,183 @@ def merge(self, update):
+    def cache(self):
+        """
+        Caches update data (stored within the doc or attributes specific in the
+        overwritten state function) to redis within a hash named for the
+        resource_id. Any attributes storing values in a dictionary are keyed by
+        the [attribute_name].[key_in_dictionary]. To be used in situation where 
+        field values are being incremented or decremented.  Otherwise, overwrite 
+        to suite your specific needs.
+        """
+        for field, data in (self.__get_update_data() or {}).items():
+            if type(data) == dict:
+                for k, v in data.items():
+                    self.process().client.hincrby(self.resource_id, "{}.{}".format(field, k), int(v))
+            else:
+                self.process().client.hincrby(self.resource_id, field, int(data))
+
+    def merge(self, other):
+        """
+        Handles how to extract a resource's data from redis and calls the user
+        defined `merge` method. To be used in situation where  field values are
+        being incremented or decremented.  Otherwise, overwrite to suite your 
+        specific needs.
+        """
+        for field in self.__get_update_data().keys():
+            self_field = getattr(self, field)
+            other_field = getattr(other, field)
+            if type(self_field) == dict:
+                for k, v in (other_field or {}).items():
+                    if k not in self_field:
+                        self_field[k] = int(v)
+                    else:
+                        self_field[k] += int(v)
+            elif self_field and other_field:
+                setattr(self, field, int(self_field) + int(other_field))
+            elif other_field:
+                setattr(self, field, int(other_field))
+
+    def __merge(self):
+        """
+        Handles how to extract a resource's data from redis and calls the user
+        defined `merge` method.
+        """
+        UpdateDoc = namedtuple("UpdateDoc", self.__get_update_data().keys())
+        update_doc = UpdateDoc(**self.get_cached_data())
+        self.merge(update_doc)
+
+    def __cache(self, block=None):
+        """
+        Handles deciding whether or not to get the resource from redis. Also
+        implements merging cached records with this one (by implementing your
+        `merge` method). Increments the number of times this record was
+        modified if the cache method executes successfully (does not raise).
+        """
+        self.cache()
+        self._clear()
+
+    def __execute(self, block=None):
+        """
+        Handles deciding whether or not to get the resource from redis. Also
+        implements merging cached records with this one (by implementing your
+        `merge` method). Increments the number of times this record was
+        modified if the cache method executes successfully (does not raise).
+        """
+        self.__merge()
+        self.execute()
+
+    def __get_update_data(self):
```

Comment:
Every attribute of the `Update` object is "update data". A better name might be "doc and state", but I think this abstraction might make unclear exactly what is going on in the methods implementing it.